### PR TITLE
Refactor differentiators

### LINKED ISSUE #299: User-facing call to parameter shift differentiator
Our use case involves using a Sample layer to obtain bitstrings, then computing a loss as a function of these bitstrings.  We wish to obtain gradients of the loss with respect to the parameters of the circuit given to the Sample layer; we want to use the Parameter Shift rule to calculate these gradients.  What would it take to break out the parameter shift code into a user-facing call?

We have example code for how we would like to use Sample layer with gradients:
```python
def energy(logits, samples):  # energy of a bernoulli with logit param
  samples = tf.cast(samples, dtype=tf.float32)
  return tf.reduce_sum(tf.nn.sigmoid_cross_entropy_with_logits(samples,
                                                               logits))
  

def sample(phi, n):
  circuit = cirq.Circuit(cirq.rx(sympy.Symbol('rx'))(cirq.GridQubit(0, 0)))
  samples = tfq.layers.Sample()(tfq.convert_to_tensor([circuit]),
                      symbol_names=['rx'],
                      symbol_values=tf.expand_dims(phi, 0),
                      repetitions=n.numpy())  # <- hacked
  return tf.cast(samples, tf.bool)


@tf.custom_gradient
def loss(phis, num_samples):
  def grad(dphis):
    # TODO!!! Parameter shift gradient
    return None, None
  return bare_loss(phis, num_samples), grad


def bare_loss(phis, num_samples):
    samples = sample(phis, num_samples)
    energies = tf.TensorArray(tf.dtypes.float32, num_samples)
    for i in tf.range(num_samples):
      this_energy = energy(tf.constant([5.0]), samples[0][i])
      energies = energies.write(i, tf.expand_dims(this_energy, 0))
    energies = energies.stack()
    return tf.reduce_mean(energies)

test_phi = tf.constant([7.0])
num_samples = tf.constant(3)

with tf.GradientTape() as g:
  g.watch(test_phi)
  loss_val = loss(test_phi, num_samples)

phi_grad = g.gradient(loss_val, test_phi)
print(loss_val)
print("phi grad: {}".format(phi_grad))
```

We have a notebook we started to prototype solutions: https://colab.research.google.com/drive/1h6aUeQ6rLVhd0sE600kA0X6hDSAazw4o#scrollTo=ZXpGyhYSNpFK

### PR DESCRIPTION
**Pull request by zaqqwerty**
Puts differentiator logic into its own function to expose intermediate circuits to the user.  Resolves #299

### CODE REVIEW

File: docs/tutorials/gradients.ipynb [hunk 1]
```diff
@@ -503,19 +519,19 @@
...
-      "outputs": [],
       "source": [
         "test_value = 0.\n",
         "\n",
         "print('Finite difference:', my_grad(pauli_z, test_value))\n",
         "print('Sin formula:      ', -np.pi * np.sin(np.pi * test_value))"
-      ]
+      ],
+      "execution_count": null,
+      "outputs": []
```
Comment (MichaelBroughton) [inline docs/tutorials/gradients.ipynb:534]: Once you get the notebook working, don't forget to restart it and clear the outputs.

File: tensorflow_quantum/python/differentiators/differentiator.py [hunk 2]
```diff
@@ -179,6 +179,27 @@ def refresh(self):
             del self.expectation_op
         return self
 
+    @abc.abstractmethod
+    def get_intermediate_logic(self, programs, symbol_names, symbol_values,
+                               pauli_sums):
+        """Returns circuits and values which are used by both diff types.
```
Comment (MichaelBroughton) [inline tensorflow_quantum/python/differentiators/differentiator.py:185]: Going to need more description here if a user is going to understand how to implement this effectively.

File: tensorflow_quantum/python/differentiators/linear_combination.py [hunk 3]
```diff
@@ -96,16 +96,14 @@ def __init__(self, weights, perturbations):
...
     @tf.function
-    def differentiate_analytic(self, programs, symbol_names, symbol_values,
-                               pauli_sums, forward_pass_vals, grad):
+    def get_intermediate_logic(self, programs, symbol_names, symbol_values,
+                               pauli_sums):
+        """Returns copies of the input programs for each perturbed symbol.
 
-        # these get used a lot
-        n_symbols = tf.gather(tf.shape(symbol_names), 0)
+        See base class for Args.
```
Comment (MichaelBroughton) [inline tensorflow_quantum/python/differentiators/linear_combination.py:103]: What is being returned here ? The base class just says what the args are but intermediate logic for each differentiation algorithm could be drastically different and should be documented.

File: tensorflow_quantum/python/differentiators/parameter_shift.py [hunk 4]
```diff
@@ -102,57 +157,20 @@ def differentiate_analytic(self, programs, symbol_names, symbol_values,
...
+        total_programs = n_param_gates * n_programs * n_shifts * n_symbols
         n_tile = n_shifts * n_param_gates * n_symbols
-        flat_perturbations = tf.concat([
-            tf.reshape(
-                tf.tile(tf.expand_dims(symbol_values, 0),
-                        tf.stack([n_tile, 1, 1])), [total_programs, n_symbols]),
-            tf.expand_dims(flat_shifts, axis=1)
-        ],
-                                       axis=1)
         flat_ops = tf.reshape(
```
Comment (MichaelBroughton) [inline tensorflow_quantum/python/differentiators/parameter_shift.py:171]: This isn't being returned by `get_intermediate_logic` function even though it is essential for the expectation calculation below, we shouldn't force a user to have to think harder than they need to on how to call into the op in their own way should they want to.

File: tensorflow_quantum/python/differentiators/differentiator.py (hunk 2 above)
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py:185]: Expanded the docstring

File: tensorflow_quantum/python/differentiators/linear_combination.py (hunk 3 above)
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/linear_combination.py]: Changed the API of `get_intermediate_logic`, the base class now describes both `Args` and `Returns`

File: tensorflow_quantum/python/differentiators/parameter_shift.py (hunk 4 above)
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/parameter_shift.py:171]: Refactored

**Comment by zaqqwerty** [comment]
@MichaelBroughton I came up with a new API for the `get_intermediate_logic` function which I think provides the flexibility needed to implement SG diffs cleanly.  I implemented it fully for `LinearCombination`, but I wanted to get your thoughts on it before I go ahead with the unit tests and implementation for PS.   Also, I forked off the initial removal of SGD stuff into its own PR #372 so that this one doesn't get too cluttered.

File: tensorflow_quantum/python/differentiators/differentiator.py [hunk 5]
```diff
@@ -179,6 +179,73 @@ def refresh(self):
...
+                The first dimension is the length of the input `programs`, the
+                second dimension is the length of the second dimension of the
+                input `pauli_sum`, the third dimension is the length of the
+                input `symbol_names`, and the last two dimensions are the same
+                shape as the output `flat_pauli_sums`.  The value at index
+                `ijkmn` is the amount of weight to give the measurement result
+                at `mn` when summing over expectation values to obtain the
+                gradient of the symbol `k` with respect to pauli sum `j` of
+                program `i`.
+        """
```
Comment (MichaelBroughton) [inline tensorflow_quantum/python/differentiators/differentiator.py:247]: Given the potentially drastically different "intermediate logic blocks" between differentiators. I think it might make more sense to just have the return signatures defined on a per differentiator basis and leave the parent class as an abstract method. Trying to fit all differentiators under this roof might prove pretty hard.

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
Looking good, a major change we should adopt here is letting the differentiator define it's own return signature, as we introduce more it will be harder and harder to keep a standardized return signature. An advanced user would also likely know that and expect different forms as well.

File: tensorflow_quantum/python/differentiators/differentiator.py (hunk 5 above)
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py:247]: Even though every differentiator can be defined via this interface, I understand your point that it could be a bit hard to reason about.  I think it could make sense to restrict the interface slightly to assume we will use the same gradient procedure for every circuit; then the mapper tensor would be the same for every circuit in the input batch, allowing us to drop an index and reduce the size of the mapper.  I'll think a bit more about how to simplify this interface along that line.

Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py]: Changed to a new, simpler batch interface.  ParameterShift and LinearCombination have been updated to obey the interface.  I still need to write tests for `get_intermediate_logic` confirming shapes and values explicitly.

File: docs/tutorials/gradients.ipynb (hunk 1 above)
Comment (zaqqwerty) [reply docs/tutorials/gradients.ipynb:534]: This is still how it looks after going to "Runtime > Restart runtime" and then doing "Edit > Clear all outputs" before downloading

File: tensorflow_quantum/python/differentiators/differentiator.py (hunk 5 above)
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py]: Now added an explicit equation at the end of the docstring to justify why the returns have the shapes specified.

**Comment by zaqqwerty** [comment]
Given your comment on the interface difficulty, I simplified things a bit to a batch API, where the same transformation is applied to each circuit in the input.  I refactored `LinearCombination` and `ParameterShift` into compliance with this new interface.  I added an explicit equation at the end of the base class docstring to help the user understand the shapes of the returns and how to use them.  Finally, I added annotated tests for these new `get_intermediate_logic` functions.

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
The interface change seems reasonable although I personally still think it is still pretty complicated.

A more major problem that I have with this re-implementation is that we have moved from making just ONE call to the provided expectation op to many seperate calls to the given expectation function inside of a `map_fn`. Looking at the docs page for `map_fn` we see: 

> map_fn unstacks elems on axis 0 to obtain a sequence of elements; calls fn to transform each element; and then stacks the transformed values back together.

In our case, this will drastically degrade performance both in simulation and on a real QPU. Simulation with have to allocate / simulate and deallocate memory for each individual circuit which is very wasteful. If the circuits are small it will also be unable to parallelize very much ( if at all ). When running on QPU the performance will also be worse. If we are unable to see the ENTIRE batch of circuits we need torun , we will never be able totake full advantage of more efficient engine methods like `run_batch` (from Cirq) or `run_sweep` (from Cirq).

If we want to keep this structure I think we should definitely fix this issue inside of the `differentiate_***` methods too.

**Comment by zaqqwerty** [comment]
Sure, I can just stack the circuits first instead

**Comment by zaqqwerty** [comment]
Reshaped things in the default diff logic so that expectation op is only called once per differentiation again

**Comment by zaqqwerty** [comment]
Had to double the number of samples in the new gradient tests to prevent flaking

**Review by MichaelBroughton — COMMENTED** [review]
I'm sorry that this has turned into such a long discussion. Perhaps it is a good indicator that we might need a design document....

Now I'm wondering how get_intermediate_logic might behave with something like the SGDifferentiator where the number of samples taken matters a great deal. The `get_intermediate_logic` method does not take num_samples as an argument which might be something we want to include ?

Would you be open to discussing this more in tomorrow meeting ?

**Comment by zaqqwerty** [comment]
Sure, let's chat about it tomorrow, I want to make sure it works for everyone's use cases.  I spoke with Jae about this new interface, he was of the opinion it will ease the re-implementation of the SGDifferentiators (those also used just a simple tiling of the input num_samples, but we can discuss including it as an argument to `get_intermediate_logic`)

**Comment by zaqqwerty** [comment]
I thought more about your comment on `num_samples`.  I was able to include it as part of the input and output of `get_intermediate_logic`, and this actually allowed the definitions of `differentiate_analytic` and `differentiate_sampled` to be pulled all the way to the base class only, greatly reducing code duplication.

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
Looking at the code I remember how tricky it was implementing these things the first time. Would we be open to doing this changeover in stages:

1. Add the interface and functionality of `get_intermediate_logic`. (This PR)
2. Move over all of the `differentiate_analytic` and `differentiate_sampled` to use them (another PR) ?

File: tensorflow_quantum/python/differentiators/differentiator.py [hunk 6]
```diff
@@ -180,14 +180,111 @@ def refresh(self):
...
+
+            As an explicit equation, the return values are defined in terms of
+            the argument values such that:
+
+            d(<programs[i](symbol_values[i])|
+                pauli_sums[i, j]|programs[i](symbol_values[i])>)
+            / d(symbol_names[k]) = sum_mn (batch_mapper[i, j, k, m, n] *
+                <batch_programs[i][m](batch_symbol_values[i][m])|
+                    batch_pauli_sums[i, m, n]
+                        |batch_programs[i][m](batch_symbol_values[i][m])>)
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py:277]: Here `d` is the differentiation symbol

File: tensorflow_quantum/python/differentiators/differentiator.py [hunk 7]
```diff
@@ -180,14 +180,111 @@ def refresh(self):
         return self
 
     @abc.abstractmethod
+    def get_intermediate_logic(self, programs, symbol_names, symbol_values,
+                               pauli_sums, num_samples):
+        """Returns components that can be used to build the gradient.
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py:185]: Slightly too long for first line only, added shortened version

File: tensorflow_quantum/python/differentiators/differentiator.py [hunk 8]
```diff
@@ -180,14 +180,111 @@ def refresh(self):
...
+        the input batch will have the same gradient procedure applied to it.
+        Thus the outer dimension of all returned components is the same as the
+        input `programs` batch dimension.
+
+        For each entry in the batch dimension of the returned values,
+        inheriting differentiators are free to generate new circuits, symbols,
+        and measurement ops as needed. To combine these outputs into the
+        required derivatives, the last return value is a map specifying the
+        linear combination of batched expectation values to use to calculate the
+        derivative of each input parameter.
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py:207]: Added a `NOTE:` paragraph pointing this out

File: tensorflow_quantum/python/differentiators/differentiator.py [hunk 9]
```diff
@@ -180,14 +180,111 @@ def refresh(self):
...
+                `symbol_values` should be placed inside of the circuits in
+                `programs`.
+            symbol_values: `tf.Tensor` of real numbers with shape
+                [batch_size, n_params] specifying parameter values to resolve
+                into the circuits specified by programs, following the ordering
+                dictated by `symbol_names`.
+            pauli_sums: `tf.Tensor` of strings with shape [batch_size, n_ops]
+                containing the string representation of the operators that will
+                be used on all of the circuits in the expectation calculations.
+            num_samples: `tf.Tensor` of positive integers representing the
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py:223]: Added the shape

[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/differentiator.py:223]: Makes sense to make this optional; do you know of any best practices around optional arguments to functions with `tf.function`?  My experience is these can get finicky

**Comment by zaqqwerty** [comment]
Hi @MichaelBroughton, I addressed your comments on the docstring.  I also broke this PR in two as you suggested; this one just adds the `get_intermediate_logic` interface, implementations, and tests; #379 deals with removing the old `differentiate_analytic` and `differentiate_sampled` code.

File: tensorflow_quantum/python/differentiators/differentiator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/python/differentiators/differentiator.py:231]: I'm pretty sure tf.functions don't support the intuitive python style default arguments, but an easy workaround is to just have a null value tensor that indicates the `get_intermeidate_logic` function won't process the samples tensor. For example you could pass in tf.constant(-2) with shape = `()` to indicate you don't need samples.

File: tensorflow_quantum/python/differentiators/differentiator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/python/differentiators/differentiator.py:285]: Might want to clear that up so it's easy to understand in the doc renderings.

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
Tested this locally on my machine and made a worrying discovery: In a relatively moderate setting of a batch of 64 circuits each with 25 symbols , 10 ops and a central-difference differentiatior, just calling `get_intermediate_logic` consumed a huge amount of memory. I worked out the numbers and I think the `batch_mapper` is the culprit with shape [64, 10, 25, 50, 50] =>  40 million entries (the last two might be 25's ?) which is far too many. 

I think we need to reconsider this component. Is there any way we could shrink down this tensor and maintain fully intuitive generality ? Maybe this makes a case for moving to each differentiator having it's own unique return signature on `get_intermediate_logic` ?

**Comment by zaqqwerty** [comment]
@MichaelBroughton I think you're right that the discussion would be aided by a design doc.  I've started writing one (sent you a DM)

**Comment by zaqqwerty** [comment]
After some more thought, since this PR just adds the functionality of intermediate logic and doesn't remove the way it functions elsewhere, could we put this in as-is (after the docstring and optional `num_samples` arg updates), then open an issue and do the careful design doc before swapping this logic into `differentiate_analytic` and `differentiate_sampled`?

To follow up on the memory issue, in your example, `batch_mapper` would actually have shape [64, 10, 25, 51, 10] which implies 8,160,000 entries.  This is about a factor of 100 larger than the existing intermediate steps of LinearCombination [here](https://github.com/tensorflow/quantum/blob/0712199eed1ce77a3b736d886c757ea48d41987f/tensorflow_quantum/python/differentiators/linear_combination.py#L152), `all_perturbations`, which for your example would have shape `[n_symbols, n_perturbations, n_programs, n_symbols]` ==> [25, 2, 64, 25] which is 80,000 entries.  If this is too large, one way that comes to mind to reduce the memory footprint might be to switch batch mapper to a [sparse tensors](https://www.tensorflow.org/api_docs/python/tf/sparse/SparseTensor), since most of the entries in the batch mapper (and in `all_perturbations`) in this case are zero.

**Comment by MichaelBroughton** [comment]
I don't think taking on a factor 100x memory growth and then moving to make `differentiate_analytic` and `differentiate_sampled` depend on them would be a wise move here. Let's take a look at the design document and then circle back once we've found some solutions that work on all fronts.

**Comment by zaqqwerty** [comment]
Completed design doc has a slightly different proposed interface, closing this one
